### PR TITLE
Support export/import states with arbitrary players.

### PR DESCRIPTION
**Pull request by asvitkine**
Previously, the dump/setup game state dev options were only able to handle a 2-player human vs. AI game. This PR expands the functionality to support any number of players and ignores the type of player (naming them p0,p1,...,pN). The previous names of ai and human are supported for backwards compatibility.

### CODE REVIEW

**Comment by kevlahnota** [comment]
Didn't test but IIRC the puzzle uses gamestate too, does it work?

**Comment by asvitkine** [comment]
> Didn't test but IIRC the puzzle uses gamestate too, does it work with the existing puzzle mode?

Yes it does. :)

**Review by tool4ever — CHANGES_REQUESTED** [review]
Attachments broken:
> ERROR: Tried to create a non-existent card named Curse of BloodlettingP1 (set: any) when loading game state!

File: forge-ai/src/main/java/forge/ai/GameState.java [hunk 1]
```diff
@@ -712,10 +621,11 @@ protected void applyGameOnThread(final Game game) {
 
         // Set negative or zero life after state effects if need be, important for some puzzles that rely on
         // pre-setting negative life (e.g. PS_NEO4).
-        if (humanLife <= 0) {
-            human.setLife(humanLife, null);
-        } else if (computerLife <= 0) {
-            ai.setLife(computerLife, null);
+        for (int i = 0; i < playerStates.size(); i++) {
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/GameState.java:624]: Just use for-each here and following further below?

**Comment by asvitkine** [comment]
> Attachments broken:
> 
> > ERROR: Tried to create a non-existent card named Curse of BloodlettingP1 (set: any) when loading game state!

Thanks, fixed! A line of code was accidentally removed that caused this. Should be good now.

**Comment by asvitkine** [comment]
On Mon, Dec 26, 2022 at 5:22 AM tool4ever ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In forge-ai/src/main/java/forge/ai/GameState.java
> <https://github.com/Card-Forge/forge/pull/2168#discussion_r1057218259>:
>
> > @@ -712,10 +621,11 @@ protected void applyGameOnThread(final Game game) {
>
>          // Set negative or zero life after state effects if need be, important for some puzzles that rely on
>          // pre-setting negative life (e.g. PS_NEO4).
> -        if (humanLife <= 0) {
> -            human.setLife(humanLife, null);
> -        } else if (computerLife <= 0) {
> -            ai.setLife(computerLife, null);
> +        for (int i = 0; i < playerStates.size(); i++) {
>
> Just use for-each here and following further below?
>
We index into two different player lists, so can't.

**Review by tool4ever — APPROVED** [review]
nice work

**Pull request merged**